File: LEETCODE/PYTHON/BuySell.py
# 121. Best Time to Buy and Sell Stock - Level(Easy) - Topic(Array, Dynamic Programming)

def BestTime(prices):
    buyCost = prices[0]
    maxProfit = 0

    for idx in range(len(prices)):
        if buyCost < prices[idx]:
            currProfit = prices[idx] - buyCost
            maxProfit = max(maxProfit, currProfit)
        else:
            buyCost = prices[idx]

    return maxProfit


prices = [7,1,5,3,6,4]
print(BestTime(prices))

File: LEETCODE/JAVA/BuySell.java
// 121. Best Time to Buy and Sell Stock - Level(Easy) - Topic(Array, Dynamic Programming)

public class BuySell {
    public static int BestTime(int[] prices){
        int buyCost = prices[0];
        int maxProfit = 0;
        for(int idx = 1; idx < prices.length; idx++){
            if(buyCost < prices[idx]){
                int currProfit = prices[idx] - buyCost;
                maxProfit = Math.max(maxProfit, currProfit);
            }else{
                buyCost = prices[idx];
            }
        }
    return maxProfit;
    }

    public static void main(String[] args) {
        int[] prices = {7,1,5,3,6,4};
        System.out.println(BestTime(prices));
    }
}
